Branch with the meson build-system

tmerrick1 opened the pull request:
The meson/ninja build system.

adamjstewart commented:
This looks good to me, although I wonder if all of the complexity borrowed from `CMakePackage` is necessary in this case. But I don't really know Meson all that well.

Would you be willing to write documentation along the lines of #5015? You can either add it to this PR and wait for #5015 to be merged, or contribute it directly to #5015.

tmerrick1 commented:
Added to create and modified build_environment for dual build system.

Meson corresponds to cmake and ninja to make that is why I copied from the cmake files.

adamjstewart commented:
I just remembered one other thing that needs to be updated. Can you add a line to `lib/spack/spack/test/build_system_guess.py` for `meson`? It looks like `autoreconf` and `octave` are also missing from the test, so you can add them if you want.

scheibelp commented:
@tmerrick1 were you still working on this? There are a couple remaining review comments. I'm assuming that https://github.com/spack/spack/pull/8467#discussion_r204257048 is addressed, but you may want to address https://github.com/spack/spack/pull/8467#discussion_r204256997.

adamjstewart commented on lib/spack/docs/build_systems/mesonpackage.rst:
```diff
@@ -0,0 +1,89 @@
+
+Any of these phases can be overridden in your package as necessary.
+There is also a ``check`` method that looks for a ``check`` target
+in the Makefile. If a ``check`` target exists and the user runs:
```
Not a Makefile

adamjstewart commented on lib/spack/spack/build_environment.py:
```diff
@@ -553,6 +556,22 @@ def get_std_cmake_args(pkg):
     return spack.build_systems.cmake.CMakePackage._std_args(pkg)
 
 
+def get_std_meson_args(pkg):
```
I think we _might_ want to avoid this. The CMake equivalent was added back in the day before separate build systems existed, and was only retained because there are still many CMake packages that haven't been converted to the `CMakePackage` build system. The only reason we would want this is if we had a package that used Meson for some versions and another build system for others.

tmerrick1 replied:
This was added late due to having to use a Package for an application (atk) that is converting from AutotoolsPackage to MesonPackage.  If we want to maintain history, then we have to do both.

adamjstewart commented on lib/spack/spack/build_systems/meson.py:
```diff
@@ -0,0 +1,185 @@
+        return [os.path.join(self.build_directory, 'meson-logs/meson-log.txt')]
+
+    @property
+    def root_mesonlists_dir(self):
```
Does Meson do the thing that CMake does where you have recursive CMakeLists.txt files? If that's the case, then this is fine, but I would rename it to `root_meson_build_dir`. If not, I would rename this `meson_build_dir` or `mesonbuild_directory` to be more in line with `AutotoolsPackage`. The recursion is the only reason it was called `root`, to distinguish that there can be many directories containing the file.

tmerrick1 replied:
It allows subprojects.  I have not seen it yet, but this build system is new.

adamjstewart commented on lib/spack/spack/build_systems/meson.py:
```diff
@@ -0,0 +1,185 @@
+
+    @property
+    def root_mesonlists_dir(self):
+        """The relative path to the directory containing meson-build
```
`meson.build`, not `meson-build`.

adamjstewart commented on lib/spack/spack/build_systems/meson.py:
```diff
@@ -0,0 +1,185 @@
+        meson, except:
+
+        * ``--prefix``
+        * ``--backend``
```
Is `--backend` actually being passed?

adamjstewart commented:
Oh, you should also update `lib/spack/spack/cmd/configure.py` and `lib/spack/spack/cmd/build.py`.

adamjstewart commented on lib/spack/docs/build_systems/mesonpackage.rst:
```diff
@@ -0,0 +1,85 @@
+------------
+
+Much like Autotools and CMake, Meson is a build system.  But it is
+meant to be both fast and user friendly as possible.  GNOME's goal
```
I would say "_as_ user friendly as possible".

adamjstewart commented on lib/spack/docs/build_systems/mesonpackage.rst:
```diff
@@ -0,0 +1,85 @@
+Build system dependencies
+^^^^^^^^^^^^^^^^^^^^^^^^^
+
+At the bare minimum, packages that use the Meson build system needs
```
typo: needs -> need
